File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxAncestorDiff(TreeNode root) {
         HashMap<Integer, LinkedList<TreeNode>> nodeAndAncestors = new HashMap<>();
        int max = 0;
        LinkedList<TreeNode> ancestors = new LinkedList<>();
        ancestors.add(root);
        nodeAndAncestors.put(root.val, ancestors);
        LinkedList<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            LinkedList<TreeNode> nodeAncestors = nodeAndAncestors.get(node.val);
            if (node.left != null) {
                LinkedList<TreeNode> leftAncestors = new LinkedList<>(nodeAncestors);
                leftAncestors.add(node.left);
                nodeAndAncestors.put(node.left.val, leftAncestors);
                queue.add(node.left);
            }
            if (node.right != null) {
                LinkedList<TreeNode> rightAncestors = new LinkedList<>(nodeAncestors);
                rightAncestors.add(node.right);
                nodeAndAncestors.put(node.right.val, rightAncestors);
                queue.add(node.right);
            }
        }
        for (LinkedList<TreeNode> ancestorList : nodeAndAncestors.values()) {
            int min = Integer.MAX_VALUE;
            int max2 = Integer.MIN_VALUE;
            for (TreeNode ancestor : ancestorList) {
                if (ancestor.val < min) min = ancestor.val;
                if (ancestor.val > max2) max2 = ancestor.val;
            }
            int diff = max2 - min;
            if (diff > max) max = diff;
        }
        return max;
    }
}

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        leafs = []
        parents = dict()
        q = deque()
        q.append((root, None))
        
        while q:
            node, parent = q.popleft()
            
            parents[node] = parent
            if not node.left and not node.right:
                leafs.append(node)
            
            if node.left:
                q.append((node.left, node))
                
            if node.right:
                q.append((node.right, node))
                
        
        ans = float('-inf')
        
        for node in leafs:
            maxAtPath = float('-inf')
            minAtPath = float('inf')
            
            p = node
            while p:
                maxAtPath= max(maxAtPath, p.val)
                minAtPath= min(minAtPath, p.val)
                p = parents[p]
            
            ans = max(ans, abs(minAtPath- maxAtPath))
        
    
        return ans
        